### PR TITLE
test(infrastructure): add Go output contract tests for root Terraform module

### LINKED ISSUE #290: test(infrastructure): Go/Terratest contract and integration tests — tracking issue
## Summary

Adopt Go/Terratest for contract validation and integration testing of the root Terraform deployment. Contract tests validate output declarations at zero cost using `terraform-docs json`; deployment tests run full infrastructure lifecycle validation on a schedule.

## Context

This complements native `terraform test` (Tier 1, module-level) with integration-level validation. The microsoft/edge-ai repository uses this exact two-tier pattern. Contract tests validate that the root module's output declarations match a strongly-typed Go struct, catching output contract breakage at PR time with zero Azure credentials and sub-second execution.

## Goals

| ID | Statement | Priority |
|----|-----------|----------|
| G-001 | Add Go toolchain to DevContainer | P0 |
| G-002 | Create Go module with shared test utilities | P0 |
| G-003 | Implement contract tests validating output declarations | P0 |
| G-004 | Integrate contract tests into CI workflow | P1 |
| G-005 | Create deployment test framework for nightly validation | P2 |

## Test Directory Structure

```text
infrastructure/terraform/tests/
├── go.mod, go.sum
├── testutil/
│   ├── contract.go              # terraform-docs JSON parsing + output validation
│   └── deployment.go            # DeployTerraform(), DestroyTerraform() wrappers
├── outputs.go                   # InfraOutputs struct with `output` tags
├── contract_test.go             # TestTerraformOutputsContract (zero cost, <1s)
├── deploy_test.go               # TestTerraformInfrastructureDeploy (nightly only)
├── run-contract-tests.sh        # Developer helper script
└── run-deployment-tests.sh      # Developer helper script
```

### Foundation

- [ ] Add Go devcontainer feature (`ghcr.io/devcontainers/features/go:1`, version 1.22+)
- [ ] Create `infrastructure/terraform/tests/` Go module with `testutil/` package
- [ ] Implement `testutil/contract.go` for `terraform-docs json` output parsing

### Contract Tests

- [ ] Define `InfraOutputs` struct matching root module output declarations
- [ ] Implement `contract_test.go` using reflection-based output key validation
- [ ] Create `run-contract-tests.sh` developer helper script

### CI Integration

- [ ] Add contract test step to CI workflow (zero auth, sub-second)
- [ ] Update `copilot-instructions.md` validation table with Go contract test command

### Integration Tests (deferred — requires Azure credentials in CI)

- [ ] Create deployment test framework with `test_structure` stages and `SKIP_*` support
- [ ] Configure GitHub OIDC federation for test Azure subscription
- [ ] Create scheduled nightly workflow for deployment tests

## Technology Stack

| Component | Version | Purpose |
|-----------|---------|---------|
| Go | 1.22+ | Test language |
| Terratest | v0.54.0 | Infrastructure testing library |
| testify | v1.11.1 | Test assertions |
| terraform-docs | (already installed) | Output contract extraction |

## Contract Test Pattern

```go
type InfraOutputs struct {
    PlatformOutputs map[string]any `output:"platform"`
    SilOutputs      map[string]any `output:"sil"`
}

func TestTerraformOutputsContract(t *testing.T) {
    var infra InfraOutputs
    keys := testutil.ExtractOutputKeys(infra)
    testutil.ValidateTerraformContract(t, "..", keys)
}
```

## Inner Dev Loop

```bash
# Contract tests (zero auth, <1 second)
cd infrastructure/terraform/tests
./run-contract-tests.sh

# Deployment tests (iterative — deploy once, re-validate many times)
SKIP_destroy=true go test -v -run TestMainDeployment -timeout 45m
SKIP_init=true SKIP_apply=true SKIP_destroy=true go test -v -timeout 5m
```

## Testing Pyramid Position

| Layer | Tool | Auth | Cost | This Issue |
|-------|------|------|------|------------|
| L0-L3 | fmt, lint, validate, test | No | $0 | — |
| **L4: Contract** | **Go contract test** | **No** | **$0** | **This issue** |
| L5: Plan | `terraform plan` | Yes | $0 | Future |
| **L6: E2E** | **Terratest (apply)** | **Yes** | **$$$** | **This issue (deferred)** |

## Dependencies

- terraform-docs (#185) — already installed in DevContainer; contract tests use `terraform-docs json` independently
- Go devcontainer feature — prerequisite for all Go tests

## References

- microsoft/edge-ai two-tier testing architecture (components + blueprints)
- Terratest documentation (gruntwork-io/terratest, 7.9k stars)
- HashiCorp `terraform test` documentation

### LINKED ISSUE #335: ci(security): add Go to CodeQL language matrix
## Summary

Add `'go'` to the CodeQL language matrix in `.github/workflows/codeql-analysis.yml`.

```yaml
# Before
language: [ 'javascript', 'python' ]

# After
language: [ 'go', 'javascript', 'python' ]
```

**Priority:** P2 — defer until Go code exceeds ~500 lines. CodeQL static analysis only adds value once there is Go source code with meaningful call graphs to scan.

**Depends on:** #290 implementation PR with Terratest code must be merged first. Do not implement this issue until Go source code exists in the repository.

## Acceptance Criteria

- [ ] CodeQL runs Go analysis alongside JavaScript and Python
- [ ] No false-positive security findings on initial Go codebase

### PR DESCRIPTION
**Pull request by bindsi**
This PR completes the final sub-issue of the Go/Terratest umbrella (#290) by adding a fast, credential-free contract test that catches drift between the root Terraform module's declared `output` blocks and the typed `InfraOutputs` struct end-to-end tests rely on. The test runs in under one second, requires no `terraform init` and no Azure credentials, and uses *terraform-docs* to extract declared outputs at test time.

It also closes #335 by adding `go` to the CodeQL language matrix so the new Go code is scanned alongside JavaScript/TypeScript and Python from day one.

## Description

The new contract pattern lives under *infrastructure/terraform/e2e/* and has three layers. A reusable **testutil** package shells out to `terraform-docs json` to discover declared outputs, then validates them against a required-key set produced by reflection over `output:"..."` struct tags. A typed *InfraOutputs* struct in *outputs.go* defines that contract for the root module's 27 outputs. A thin `TestTerraformOutputsContract` entry point wires the two together and is the first real test in the previously placeholder `e2e` package.

CI wiring lives in *.github/workflows/go-tests.yml*: a pinned `Install terraform-docs` step (v0.21.0, SHA256 verified) runs between `Setup Go` and `Run Go tests`. A small developer helper, *run-contract-tests.sh*, provides a one-command local entry point with preflight checks for `go` and `terraform-docs`. The `*.go` row in *.github/copilot-instructions.md* now cites the helper alongside the existing lint and test commands.

### Test infrastructure

The contract is intentionally **required-subset-of-declared**. Adding a new output to *outputs.tf* without updating *InfraOutputs* leaves the test green; removing a declared output that the struct still references fails with an actionable message naming *outputs.tf*. Both directions were probed during validation.

- New **testutil** package exposing `GetTerraformDeclaredOutputs`, `ValidateOutputContract`, `ValidateTerraformContract`, and `GetOutputKeysFromStruct` — reused unchanged by future per-deployment contracts.
- New `InfraOutputs` struct mirroring the root module's 27 outputs, each tagged with `output:"<name>"`; `RequiredOutputKeys()` extracts the tag set via reflection.
- New `TestTerraformOutputsContract` in *contract_test.go* — replaces the placeholder *e2e.go* file as the package's anchor.
- Unit tests in *testutil/contract_test.go* cover value/pointer receivers, empty struct, untagged struct, and contract-validation success paths so the Codecov *go* flag reports meaningful coverage on the helper.
- *go.mod* / *go.sum* now require `github.com/stretchr/testify v1.11.1`.

### CI and tooling

- New **Install terraform-docs** step in *go-tests.yml*, pinned to **v0.21.0** with explicit SHA256 verification matching the literal already used by *terraform-docs-check.yml*. Installs to `/usr/local/bin/` via `sudo mv` on `ubuntu-latest`.
- New executable *run-contract-tests.sh* with `set -euo pipefail`, a `-v`/`--verbose` flag, and clear preflight messages when `go` or `terraform-docs` is missing.
- *.github/copilot-instructions.md* `*.go` Quick Reference row updated to list the new helper and call out the `terraform-docs` dependency.
- *.github/workflows/codeql-analysis.yml* matrix gains `go` (closes #335); CodeQL autobuild discovers *infrastructure/terraform/e2e/go.mod* automatically.

### Documentation table-alignment catch-up

A pre-existing `format:tables` drift in six markdown files (*data-pipeline/arc/README.md*, *data-pipeline/setup/README.md*, *docs/data-pipeline/acsa-setup.md*, *docs/security/README.md*, *docs/security/workflow-permissions.md*, *scripts/README.md*) is included as whitespace-only edits, balanced 101 insertions / 101 deletions. No content change.

## Type of Change

- [x] ✨ New feature (non-breaking change adding functionality)
- [x] 📚 Documentation update
- [ ] 🐛 Bug fix (non-breaking change fixing an issue)
- [ ] 💥 Breaking change (fix or feature causing existing functionality to change)
- [ ] 🏗️ Infrastructure change (Terraform/IaC)
- [ ] ♻️ Refactoring (no functional changes)

## Component(s) Affected

- [ ] `infrastructure/terraform/prerequisites/` - Azure subscription setup
- [x] `infrastructure/terraform/` - Terraform infrastructure
- [ ] `infrastructure/setup/` - OSMO control plane / Helm
- [ ] `workflows/` - Training and evaluation workflows
- [ ] `training/` - Training pipelines and scripts
- [x] `docs/` - Documentation

## Testing Performed

- [ ] Terraform `plan` reviewed (no unexpected changes)
- [ ] Terraform `apply` tested in dev environment
- [ ] Training scripts tested locally with Isaac Sim
- [ ] OSMO workflow submitted successfully
- [ ] Smoke tests passed (`smoke_test_azure.py`)

Additional validation specific to this PR (all passed locally):

- `npm run test:go` — 10/10 tests pass across `e2e` and `testutil` packages
- `./infrastructure/terraform/e2e/run-contract-tests.sh` — end-to-end helper exercise green in <1s
- `npm run lint:yaml`, `npm run lint:md`, `npm run spell-check` — all clean
- `shellcheck infrastructure/terraform/e2e/run-contract-tests.sh` — clean
- Contract drift probes — extra declared outputs tolerated; missing declared output fails with the expected *outputs.tf* pointer

## Documentation Impact

- [ ] No documentation changes needed
- [x] Documentation updated in this PR
- [ ] Documentation issue filed

## Bug Fix Checklist

*Not applicable — this is a feature PR adding new tests.*

- [x] Linked to issue being fixed
- [ ] Regression test included, OR
- [ ] Justification for no regression test:

## Checklist

- [x] My code follows the [project conventions](copilot-instructions.md)
- [x] Commit messages follow [conventional commit format](instructions/commit-message.instructions.md)
- [x] I have performed a self-review
- [x] Documentation impact assessed above
- [x] No new linting warnings introduced

## Related Issues

Closes #290
Closes #335 (CodeQL Go language matrix)

## Notes

The single commit on this branch has a malformed message (no newlines between body bullets, so the entire body collapses into the subject line). Recommend amending to the conventional-commit format before push:

```text
test(infrastructure): add Go contract tests for root Terraform outputs (#290)

* add testutil package with terraform-docs-driven contract validation
* add InfraOutputs struct with 27 tagged fields mirroring outputs.tf
* add TestTerraformOutputsContract entry point (<1s, zero Azure creds)
* wire pinned terraform-docs install into go-tests.yml
* include whitespace-only table-alignment catch-up in 6 docs READMEs

🤖 - Generated by Copilot
```

## Follow-up

Tracked in the implementation plan's Planning Log:

- **WI-01** — G-005 deployment tests (real `terraform apply` against Azure)
- **WI-02** — Contract tests for Terraform submodules
- **WI-03** — Per-deployment contract tests (VPN, DNS, automation)
- **WI-04** — Flip *go-tests.yml* `soft-fail` to hard-fail once stable on CI
- **WI-05** — Composite action for `Install terraform-docs` to dedupe with *terraform-docs-check.yml*
- **WI-06 / #335** — Add `go` to the CodeQL language matrix once Go LOC exceeds ~500

### CODE REVIEW

File: infrastructure/terraform/e2e/testutil/contract_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	t.Run("extra declared outputs are allowed", func(t *testing.T) {
+		declared := []string{"a", "b", "c", "extra"}
+		required := []string{"a", "b", "c"}
+		ValidateOutputContract(t, declared, required)
+	})
+
+	t.Run("empty required set", func(t *testing.T) {
+		ValidateOutputContract(t, []string{"a", "b"}, nil)
+	})
+}
```
Comment (katriendg) [inline infrastructure/terraform/e2e/testutil/contract_test.go:68]: 🔴 **Coverage gap — failure paths untested (58.3% vs 80–100% target)**

The `go` Codecov flag was added during pre-work specifically to enforce meaningful coverage. Current tests only exercise happy paths:

- `ValidateOutputContract` (78.6%) — the `missing > 0` failure branch is never hit
- `GetTerraformDeclaredOutputs` and `ValidateTerraformContract` — 0% in `testutil` package

Suggested additions:
1. A test with missing required outputs to exercise the failure/error branch
2. An exact-match test (declared == required, no extras)
3. A `GetTerraformDeclaredOutputs` test using a `testdata/minimal-module/` fixture directory with a minimal `outputs.tf` — this also covers `ValidateTerraformContract`

File: .github/copilot-instructions.md [hunk 2]
```diff
@@ -399,7 +399,7 @@ Run `npm install` (or `npm ci`) before any `npm run` lint commands. `shellcheck`
 | `*.md` | `npm run lint:md`, `npm run spell-check`, `npm run format:tables` |
 | `*.tf`, `*.tfvars` | `npm run lint:tf`, `npm run lint:tf:validate`, `terraform plan` |
 | `*.tftest.hcl` | `npm run test:tf`, `cd infrastructure/terraform/modules/<name> && terraform test` or `cd infrastructure/terraform && terraform test` |
```
Comment (katriendg) [inline .github/copilot-instructions.md:428]: ⚠️ **`*.tf` Quick Reference row missing contract test cross-reference**

The `*.go` row was updated but contributors editing `outputs.tf` won't discover the contract requirement until CI fails. Suggest appending to the `*.tf` row:

```suggestion
| `*.tf`, `*.tfvars` | `npm run lint:tf`, `npm run lint:tf:validate`, `terraform plan`, `npm run test:go` (output contract) |
```

File: .github/workflows/go-tests.yml [hunk 3]
```diff
@@ -49,6 +49,25 @@ jobs:
           go-version-file: infrastructure/terraform/e2e/go.mod
```
Comment (katriendg) [inline .github/workflows/go-tests.yml:49]: 💡 **`cache-dependency-path` points to `go.mod` instead of `go.sum`**

Go's cache key should use `go.sum` — the lockfile that changes when dependencies update. `go.mod` only changes when direct dependency declarations change, missing transitive updates.

```suggestion
          cache-dependency-path: infrastructure/terraform/e2e/go.sum
```

**Review by katriendg — COMMENTED** [review]
This is looking fantastic — solid contract testing pattern, clean CI wiring, and the terraform-docs SHA verification is a nice touch. A few items to tighten up before merge, mostly around test coverage and documentation cross-references.

---

### Additional documentation cross-references (files not in this diff)

**⚠️ `docs/contributing/contribution-workflow.md` — Terraform modules row (line ~170)**

The `Terraform modules` validation row lists `terraform fmt`, `terraform validate`, and `terraform plan` but doesn't mention that output changes also require `npm run test:go` to validate the contract. Suggest:

```markdown
| Terraform modules | `terraform fmt`, `terraform validate`, `terraform plan` output attached to PR, `npm run test:go` (output contract) |
```

**⚠️ `docs/contributing/prerequisites.md` — Go validation section (line ~190)**

The Go block lists lint and test commands but doesn't explain that contract tests validate Terraform outputs or that `terraform-docs` is required. Suggest adding after the existing Go block:

```bash
# Contract tests (validates Terraform outputs against Go struct — requires terraform-docs)
# Run after adding/removing/renaming Terraform outputs
./infrastructure/terraform/e2e/run-contract-tests.sh
```

**💡 `docs/contributing/contribution-workflow.md` — Go modules row (line ~173)**

Contributors who run `npm run test:go` without `terraform-docs` installed will get a failure with no guidance from this table. Suggest:

```markdown
| Go modules | `npm run lint:go` (golangci-lint), `npm run test:go` (`go test`, requires `terraform-docs`) |
```

**Review by rezatnoMsirhC — COMMENTED** [review]
Thanks for a well-structured PR. The contract pattern is clean, the testutil package is reusable, and the CI wiring is correct. Left two comments.

File: infrastructure/terraform/e2e/contract_test.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+// TestTerraformOutputsContract validates root module output declarations
+// against InfraOutputs. Runs in < 1s, no Azure auth, no terraform init.
+//
+// Requirements:
+//   - terraform-docs must be installed and on PATH
+//   - Valid Terraform configuration at ../
+func TestTerraformOutputsContract(t *testing.T) {
+	testutil.ValidateTerraformContract(t, "..", InfraOutputs{}.RequiredOutputKeys())
+}
```
Comment (rezatnoMsirhC) [inline infrastructure/terraform/e2e/contract_test.go:24]: The testutil fixture tests (`TestGetTerraformDeclaredOutputs`, `TestValidateTerraformContract`) both guard with `exec.LookPath("terraform-docs")` and call `t.Skip()` when the tool is absent. This test has no such guard, so running `go test ./...` on a machine without `terraform-docs` installed produces a hard failure with an opaque `require.NoError` message rather than a clean skip. Suggest adding the same guard for consistency:

```go
func TestTerraformOutputsContract(t *testing.T) {
    if _, err := exec.LookPath("terraform-docs"); err != nil {
        t.Skip("terraform-docs not installed; skipping contract test")
    }
    testutil.ValidateTerraformContract(t, "..", InfraOutputs{}.RequiredOutputKeys())
}
```

File: infrastructure/terraform/e2e/testutil/contract.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+	t.Logf("Declared outputs: %v", declared)
+	t.Logf("Required outputs: %v", required)
+
+	missing := findMissingOutputs(declared, required)
+
+	if len(missing) > 0 {
+		t.Errorf("Missing %d required outputs: %v", len(missing), missing)
+		t.Errorf("Declare these in infrastructure/terraform/outputs.tf or remove from InfraOutputs struct")
+	}
```
Comment (rezatnoMsirhC) [inline infrastructure/terraform/e2e/testutil/contract.go]: Two `t.Errorf` calls (non-fatal) followed by `require.Empty` (fatal) report the same condition. The final `"output contract validation failed"` message is less informative than the preceding errors and adds noise to test output. Suggest consolidating into a single `require.Empty` with the full actionable message:

```go
func ValidateOutputContract(t *testing.T, declared, required []string) {
    t.Helper()

    t.Logf("Declared outputs: %v", declared)
    t.Logf("Required outputs: %v", required)

    missing := findMissingOutputs(declared, required)
    require.Empty(t, missing,
        "missing %d required outputs: %v -- declare these in infrastructure/terraform/outputs.tf or remove from InfraOutputs struct",
        len(missing), missing)
}
```

**Review by WilliamBerryiii — APPROVED** [review]
